add fireworks

Lev-Shapiro opened the pull request.

Commit:
add fireworks

sydnenaal commented on src/objects/effectors/explosion/firework/firework.service.ts:
```diff
@@ -0,0 +1,93 @@
+  ) {
+    this.resize()
+
+    window.requestAnimationFrame =
```
Это зачем тут? Если хотел сделать широкую поддержку браузерами - такими костылями это не делается.
Найди полифил, и используй его. И, ради бога, не здесь.

sydnenaal commented on src/objects/effectors/explosion/firework/firework.service.ts:
```diff
@@ -0,0 +1,93 @@
+    this.canvas.height = this.height
+  }
+
+  async build(particleAmount: number, color: string, coords: Coords) {
```
Либо добавь асинхронную логику, либо удали ключевое слово `async`

sydnenaal commented on src/objects/effectors/explosion/firework/firework.service.ts:
```diff
@@ -0,0 +1,93 @@
+  async build(particleAmount: number, color: string, coords: Coords) {
+    const particles: ParticleEntity[] = []
+
+    for (var i = 0; i < particleAmount; i++) {
```
Вместо тысячи итераций:
```js
const getPracticleEntity = () => new ParticleEntity(this.particlePolicy, this.ctx, color, coords);
const particles: ParticleEntity[] = Array.from(new Array(particleAmount), getPracticleEntity);
```

sydnenaal commented on src/objects/effectors/explosion/firework/firework.service.ts:
```diff
@@ -0,0 +1,93 @@
+
+    const alive: ParticleEntity[] = []
+
+    for (var i = 0; i < particles.length; i++) {
```
Многовато телодвижений, попробуй так:
```js
for (const particle of particles) {
   if (particle.move() === ParticleStatus.Alive) {
        particle.draw()
        alive.push(particle)
    }
}
```

sydnenaal commented on src/objects/effectors/explosion/firework/firework.service.ts:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    // if firework didn't end and new render loop wasn't created => continue rendering
+    if (Object.keys(this.particles).length) {
```
У тебя `fireworks` объявлена десятью строчками выше, нафига повторно вычилсять одно и то же?

Lev-Shapiro replied:
Во время вызова функции this.continueFirework происходит чистка фейерверков в объекте так что высчитанное прежде значение может отличаться от нового. Однако я бы это все равно исправил и сделал как минимум, policy при котором станет понятно что, куда и зачем

sydnenaal commented on src/objects/effectors/explosion/firework/firework.service.ts:
```diff
@@ -0,0 +1,93 @@
+
+    const fireworks = Object.keys(this.particles)
+
+    for (var i = 0; i < fireworks.length; i++) {
```
`For..of` для кого придумали?

Commit:
readability fixes

Commit:
readability fixes

Commit:
remove useless classes

remove classes that have no logic inside and only extend something from another class

Commit:
remove multiple update to same class
